### PR TITLE
Source generator v2

### PR DESCRIPTION
**Pull request by pjmagee**
#10

### CODE REVIEW

File: sdk/Dagger.SDK/Dagger.SDK.csproj [hunk 1]
```diff
@@ -15,9 +15,9 @@
                 OutputItemType="Analyzer"
                 ReferenceOutputAssembly="false" />
     </ItemGroup>
-
+    
     <ItemGroup>
-        <AdditionalFiles Include="introspection.json"/>
+        <AdditionalFiles Include="introspection.json" />
```
Comment (pjmagee) [inline sdk/Dagger.SDK/Dagger.SDK.csproj:20]: We need to keep this somehow? Or generator doesnt run code and when we delete file, `AdditionalFiles` is removed automatically it seems. So need to manually add this back in... not sure how we handle this.

File: sdk/Dagger.SDK.SourceGenerator/Dagger.SDK.SourceGenerator/Code/CodeRenderer.cs [hunk 2]
```diff
@@ -30,7 +30,7 @@ namespace Dagger.SDK;
 
                public class Scalar
                {
-                   public string Value;
+                   public string Value { get; set; }
```
Comment (pjmagee) [inline sdk/Dagger.SDK.SourceGenerator/Dagger.SDK.SourceGenerator/Code/CodeRenderer.cs:33]: Was there reason this was field?

Comment (wingyplus) [reply sdk/Dagger.SDK.SourceGenerator/Dagger.SDK.SourceGenerator/Code/CodeRenderer.cs:33]: In GraphQL, the `id` type such as, `EnvVariableID`, `ContainerID`, etc. represents as string value. And we need to render it to type. But problem I faced is I cannot find the best way to serialize string to type directly. So I ended up by introduce this property and write a json converter to serialize scalar string into this property.

File: sdk/Dagger.SDK/Dagger.SDK.csproj (hunk 1 above)
Comment (wingyplus) [reply sdk/Dagger.SDK/Dagger.SDK.csproj:20]: I think we should keep this. The generator requires this file to generate sdk. If the generator keep silence, the SDK will completely broken due to `Dagger` class requires during call `Connect`.

File: sdk/Dagger.SDK.SourceGenerator/Dagger.SDK.SourceGenerator/Code/CodeRenderer.cs [hunk 3]
```diff
@@ -54,34 +55,39 @@ public string RenderEnum(Type type)
...
-                                          return kvPairs;
-                                      }
-                                      """;
+        var toKeyValuePairsProperties = type.InputFields.Select(field => 
+            $"""
+            kvPairs.Add(KeyValuePair.Create("{field.Name}", {RenderArgumentValue(field, asProperty: true)} as Value));
+            """);
+
+        var toKeyValuePairsMethod = 
+              $$"""
```
Comment (wingyplus) [inline sdk/Dagger.SDK.SourceGenerator/Dagger.SDK.SourceGenerator/Code/CodeRenderer.cs:82]: Is this happens by formatter or you just change the line manually?

**Comment by wingyplus** [comment]
LGTM. 👍 

Could you rebase this branch? Just push a fix on code render to main and want to make sure every work perfectly. :)

File: sdk/Dagger.SDK.SourceGenerator/Dagger.SDK.SourceGenerator/Code/CodeRenderer.cs (hunk 2 above)
Comment (pjmagee) [reply sdk/Dagger.SDK.SourceGenerator/Dagger.SDK.SourceGenerator/Code/CodeRenderer.cs:33]: But it should stay as a C# `field` or `{ get; set; }`?

File: sdk/Dagger.SDK.SourceGenerator/Dagger.SDK.SourceGenerator/Code/CodeRenderer.cs (hunk 3 above)
Comment (pjmagee) [reply sdk/Dagger.SDK.SourceGenerator/Dagger.SDK.SourceGenerator/Code/CodeRenderer.cs:82]: This was me, but actually .editorconfig didn't say anything.

**Comment by pjmagee** [comment]
> LGTM. 👍
> 
> Could you rebase this branch? Just push a fix on code render to main and want to make sure every work perfectly. :)

Done. ran test and everything pass

File: sdk/Dagger.SDK/Dagger.SDK.csproj (hunk 1 above)
Comment (pjmagee) [reply sdk/Dagger.SDK/Dagger.SDK.csproj:20]: How does this work in the world of Powershell?

1. We have source generator project

2. We have this `wrapper` project (it depend on introspection.json)

3. The generator need to generate code

4. How does package look for end user? 
- no introspection.json
- dagger develop -sdk dotnet (this this add introspection.json somehow?)

File: sdk/Dagger.SDK.SourceGenerator/Dagger.SDK.SourceGenerator/Code/CodeRenderer.cs [hunk 4]
```diff
@@ -270,12 +275,12 @@ private object RenderArgumentBuilder(Field field)
         }
 
         var optionalArgs = field.OptionalArgs();
-        if (optionalArgs.Count() > 0)
+        if (optionalArgs.Any())
         {
-            optionalArgs.Aggregate(builder, (builder, arg) =>
+            optionalArgs.Aggregate(builder, (sb, arg) =>
```
Comment (pjmagee) [inline sdk/Dagger.SDK.SourceGenerator/Dagger.SDK.SourceGenerator/Code/CodeRenderer.cs:280]: This naming was cos of warning hiding outer `builder`

File: sdk/Dagger.SDK/Dagger.SDK.csproj (hunk 1 above)
Comment (wingyplus) [reply sdk/Dagger.SDK/Dagger.SDK.csproj:20]: > How does this work in the world of Powershell?
> 
> 1. We have source generator project
> 2. We have this `wrapper` project (it depend on introspection.json)
> 3. The generator need to generate code
> 4. How does package look for end user?
> 
> * no introspection.json
> * dagger develop -sdk dotnet (this this add introspection.json somehow?)

For Dagger module, the engine will inject introspection during generate code. You can see at https://github.com/wingyplus/dagger-dotnet-sdk/pull/7/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R102.

File: sdk/Dagger.SDK.SourceGenerator/Dagger.SDK.SourceGenerator/Code/CodeRenderer.cs (hunk 2 above)
Comment (wingyplus) [reply sdk/Dagger.SDK.SourceGenerator/Dagger.SDK.SourceGenerator/Code/CodeRenderer.cs:33]: Use property is fine for me. 👍

**Comment by wingyplus** [comment]
Thank you for your help. ❤️

File: sdk/Dagger.SDK.SourceGenerator/Dagger.SDK.SourceGenerator/Code/CodeRenderer.cs (hunk 3 above)
Comment (wingyplus) [reply sdk/Dagger.SDK.SourceGenerator/Dagger.SDK.SourceGenerator/Code/CodeRenderer.cs:82]: I see.

**Pull request merged**